mycpp/control_flow_pass.py

```diff
@@ -24,19 +25,98 @@ class UnsupportedException(Exception):
         self.types = types
-        self.cfgs = {}
+        self.cfgs = collections.defaultdict(pass_state.ControlFlowGraph)
         self.current_statement_id = None
+        self.current_class_name = None
         self.current_func_node = None
         self.loop_stack = []
+        self.virtual = virtual
+        self.local_vars = local_vars
+        self.imported_names = set()  # MemberExpr -> module::Foo() or self->foo
+
+        # HACK for conditional import inside mylib.PYTHON
+        # in core/shell.py
+        self.imported_names.add('help_meta')
 
     def current_cfg(self):
         if not self.current_func_node:
             return None
 
-        return self.cfgs.get(split_py_name(self.current_func_node.fullname))
+        return self.cfgs[split_py_name(self.current_func_node.fullname)]
+
+    def resolve_callee(self, o: CallExpr) -> Optional[util.SymbolPath]:
+
+        if isinstance(o.callee, NameExpr):
+            return split_py_name(o.callee.fullname)
+
+        elif isinstance(o.callee, MemberExpr):
+            if isinstance(o.callee.expr, NameExpr):
+                is_module = (isinstance(o.callee.expr, NameExpr) and
+                             o.callee.expr.name in self.imported_names)
+                if is_module:
+                    return split_py_name(
+                        o.callee.expr.fullname) + (o.callee.name, )
+
+                elif o.callee.expr.name == 'self':
+                    assert self.current_class_name
+                    return self.current_class_name + (o.callee.name, )
+
+                else:
+                    local_type = None
+                    for name, t in self.local_vars.get(self.current_func_node,
+                                                       []):
+                        if name == o.callee.expr.name:
+                            local_type = t
+                            break
+
+                    if local_type:
+                        if isinstance(local_type, str):
+                            return split_py_name(local_type) + (
+                                o.callee.name, )
+
+                        elif isinstance(local_type, Instance):
+                            return split_py_name(
+                                local_type.type.fullname) + (o.callee.name, )
+
+                        elif isinstance(local_type, UnionType):
+                            assert len(local_type.items) == 2
+                            return split_py_name(
+                                local_type.items[0].type.fullname) + (
+                                    o.callee.expr.name, )
+
+                        else:
+                            assert not isinstance(local_type, CallableType)
+                            # primitive type or string. don't care.
+                            return None
+
+                    else:
+                        # context or exception handler. probably safe to ignore.
+                        return None
+
+            else:
```

Comment:
this also seems like it may duplicate some logic in cppgen_pass

I think we may want to pass the result of resolve_callee onto a future stages, in addition to using it in datalog

Reply:
I guess the if conditions overlap with `GetCType()` somewhat, but I think a shared function would be overkill. Or do you mean specifically the `UnionType` thing?

I can definitely stash these in a map so future passes can grab it easily, though.